### ISSUE TITLE
GPU usage is 0 for a long time

### ISSUE THREAD

**Issue by COST-97** [ReinFlow/ReinFlow#6]
Hello:
Thank you for your excellent work!

My fine-tuning script:

`base_policy_path=reinflow/log/robomimic/pretrain/transport/transport_pre_reflow_mlp_img_ta8_td20/2025-09-10_17-41-37_42/checkpoint/state_2000.pt

python script/run.py --config-dir cfg/robomimic/finetune/transport \
                        --config-name ft_ppo_reflow_mlp_img \
                        denoising_steps=1 \
                        ft_denoising_steps=1 \
                        wandb.offline_mode=True \
                        +wandb.mode=offline \
                        +wandb._service_wait=120 \
                        device=cuda:0 \
                        +sim_device=cuda:1 \
                        base_policy_path=$base_policy_path`

I'm noticing that GPU usage remains at zero for extended periods while the program is running.

<img width="644" height="375" alt="Image" src="https://github.com/user-attachments/assets/7e3adc5b-0a7c-49e5-97ae-7816744f92ae" />

The terminal currently outputs:

<img width="1522" height="817" alt="Image" src="https://github.com/user-attachments/assets/33eb5eb1-828f-4867-8909-0ab6a069df15" />

What could be the problem?
Could you give me some advice?
Thank you very much!

**Comment by ReinFlow** [comment]
Hi @COST-97, have you witnessed a positive GPU utility rate during the process? Does your GPU support EGL rendering? (I suppose so). 

It is actually kind of normal when your script does not occupy GPU utility all the time. However, you should be able to see a large GPU power consumption during the optimization procedure (computing the loss and doing backward propagation). 

Moreover, we recommend using 4 denoising steps for this transport task, because it makes the model converge faster (as what we did in the paper). Maybe you can also checkout this configuration to see if your GPU is actually accelerating the compute? 

We are happy to respond to any of your thoughts! 

Best, 
ReinFlow Team. 
September 14th, 2025.

**Comment by COST-97** [comment]
Hello:
Thank you very much for your prompt response.
I followed your suggestion and set 4 denoising steps.
The results are now:

<img width="1646" height="742" alt="Image" src="https://github.com/user-attachments/assets/d7bc205a-f778-493b-a12b-517bcb2493a4" />

itr 69 | Total Steps 10.080M | Time: 5123.755, corresponding to a Success Rate of 55.50% ± 49.70% #. 

Is there some hyperparameter setting I'm missing?
In addition, I've found that increasing the number of parallel environments (500 or 1000) makes the program extremely slow. 
Are there any ways to further speed up model interaction and training?

I look forward to your reply.

**Comment by Tonghe-Zhang** [comment]
Hi @COST-97, 
Thanks for sharing your screenshot!

I noticed that the success rate is increasing, which is a good sign, but the iteration time is super long. I wonder if other programs are running in your background, which could limit the available CPU kernels? As of the success rate, it seems that your result at step 70 matches our open-source W&B logs, posted [here](https://wandb.ai/reinflow/robomimic-transport-img-finetune?nw=nwusertonghezhang01). So I guess this is not an algorithm issue, but mainly about execution speed and computation resource management. 

Regarding scaling environment numbers, please be aware that Robomimic (Robosuite) uses MuJoCo as the simulator, which by default relies on a CPU-based physical simulation engine. This implies that the maximum number of environments you can use to accelerate your program is capped by the number of available CPU kernels on your machine. As this is not a GPU-accelerated simulator, scaling the env number to 500 or 1,000 will definitely not provide the gains you may expect in GPU simulators (e.g., IsaacGym/Sim/ManiSkill), and quite on the contrary, it will likely cause your machine to jam, due to an excessively large amount of processes running together.  We recommend adjusting your environment number based on your device's CPU availability, and the other programs on your machine that shares these resources. You can check that out with simple terminal commands.
